### ISSUE TITLE
🏹 Vector / Embedding-Based Features

### ISSUE THREAD

**Issue by xehu** [Watts-Lab/team-process-map#120]
# Overview
Several of our features require analyzing the embedding of the the utterances of members in a team. The aim of the Issue is to consolidate all these issues into a central pipeline that processes the embeddings all in one go --- so that they do not have to be separately computed each time we run the main featurizer.

In other words, similar to with the lexicons (which are pre-processed just once, and then used from the pickle file thereafter), our process will look something like the following:

1. Preprocess our sentence embeddings using `preprocess_sent_vectors.ipynb`
2. This saves all of our embeddings as CSV's into a folder called `embeddings/`, located at the top level in `feature_engine`.
3. We can pull the custom embeddings from the saved CSV's by using the variable `self.word_embedding_path` in the FeatureBuilder. In other words: we can load in the pre-computed embeddings during featurization, and we do not need to compute them again and again each time.

The infrastructure for this pipeline has already been created on the branch https://github.com/Watts-Lab/team-process-map/tree/emily/vector_pipelining. This branch is a fork off of `Shruti_dd`, so it should be compatible with all progress made so far.

# Step 1: Build Initial Versions of Features and Connect the Dots
The next step is to go ahead and start building the features that require embeddings --- Discursive Diversity, LSM, Cosine Similarity, etc. --- and use the infrastructure (`self.word_embedding_path`, `embeddings/` folder) to assist. In the process, we'll figure out what else we need to add to the infrastructure to make things work.

Notably, these features should get built and saved in `features/`, and then called only at the `calculate_conversation_level_features.py` --- we are BYPASSING the chat-level features, since we already pre-processed every chat and obtained the relevant embeddings.

This will work similarity to the way that `get_gini_features()` currently works: Gini is a feature that gets added separately to the conversation_level features, instead of "summarizing" chat-level features.

[Sample code for Gini]
```
 def get_gini_features(self) -> None:
        """
            This function is used to calculate the gini index for each conversation 
            based on the word level and character level information.
        """
        # Gini for #Words
        self.conv_data = pd.merge(
            left=self.conv_data,
            right=get_gini(self.chat_data, "num_words"),
            on=['conversation_num'],
            how="inner"
        )
        # Gini for #Characters
        self.conv_data = pd.merge(
            left=self.conv_data,
            right=get_gini(self.chat_data, "num_chars"),
            on=['conversation_num'],
            how="inner"
        )
```

## Checklist for Step 1

- [ ] Build your feature and save it in `features/`; use the pre-computed embeddings!
- [ ] Call your features in `calculate_conversation_level_features.py`
- [ ] Evaluate your feature. How "accurate" is it? Do the results match your intuitions? You may have to creatively come up with some rough metric, or do some searching at this stage. **(This is very important --- you should have a specific metric in mind before moving on to Step 2)**

# Step 2: Fine-Tune!
In Step 1, we are using SBERT vectors with no additional fine-tuning. In Step 2, I think it will be worthwhile to explore how fine-tuning the models with additional domain-specific vocabulary may help improve our performance. (This is also why it's important to have a sense of performance metrics from Step 1, as we need a way to quantify how much "better" we are doing!)

The main type of fine-tuning that we will likely benefit from is masked language modeling (https://huggingface.co/learn/nlp-course/chapter7/3?fw=tf), or MLM. In other words, we want to teach the model specific words and contexts that are relevant to a particular domain, instead of whatever was used in the original training data (some generic dataset).

This section is significantly more open-ended, but broadly, it will look like this...

- Following the sentence transformers documentation for performing MLM: https://www.sbert.net/examples/unsupervised_learning/MLM/README.html
- Code sample: https://github.com/UKPLab/sentence-transformers/blob/master/examples/unsupervised_learning/MLM/train_mlm.py
- Starting with one dataset (Juries is a good candidate, since the language is the most 'different'), implement MLM. This will likely take place in a separate .ipynb, outside of our main Featurization pipeline --- since the idea is that we'll do all this preprocessing separate from running `featurize.py`.
- Generate new embeddings and save them.
- Re-run the feature you're working on with the new embeddings.
- Do the metrics get any better after fine-tuning? (If not, maybe it's not worth it!)

**Comment by xehu** [comment]
I also found this link helpful when learning about sentence embeddings: https://www.pinecone.io/learn/sentence-embeddings/
